src/main/kotlin/org/rust/ide/intentions/NestUseStatementsIntention.kt

```diff
@@ -0,0 +1,207 @@
+        val useGroupOnCursor = element.ancestorStrict<RsUseGroup>()
+        val useSpeckOnCursor = element.ancestorStrict<RsUseSpeck>() ?: return null
+
+        return if (useGroupOnCursor != null) {
+            PathInGroup.create(useGroupOnCursor, useSpeckOnCursor)
+        } else {
+            PathInUseItem.create(useItemOnCursor, useSpeckOnCursor)
+        }
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val useSpecksByBasePath = mapSpecksByBasePath(ctx)
+
+        val paths = makeGroupedPath(useSpecksByBasePath)
+
+        val root = ctx.root()
+        var firstInserted: PsiElement? = null
+        var lastInserted: PsiElement? = null
+
+        for (path in paths) {
+            val addPoint = lastInserted ?: ctx.firstPreviousElement()
+            val added = root.addAfter(ctx.createElement(path, project), addPoint)
+            lastInserted = added
+            firstInserted = firstInserted ?: added
+        }
+
+        for (prevElement in ctx.getPreviousElements()) {
+            prevElement.delete()
+        }
+
+        editor.caretModel.moveToOffset(firstInserted!!.textRange.startOffset + ctx.cursorOffset)
+    }
+
+    private fun mapSpecksByBasePath(ctx: Context): Map<String, List<RsUseSpeck>> {
+        val useSpeckByBasePath = mutableMapOf<String, MutableList<RsUseSpeck>>()
+        for (useSpeck in ctx.useSpecks) {
+            val path = useSpeck.path
+            val group = useSpeck.useGroup
+
+            val useSpeckStartsWithGroup = path == null && group != null
+            if (useSpeckStartsWithGroup) {
+                useSpeck.forEachLeafSpeck {
+                    if (it.path != null) {
+                        useSpeckByBasePath.getOrPut(it.path!!.basePath().referenceName) { mutableListOf() }.add(it)
+                    }
+                }
+            } else {
+                if (path != null) {
+                    useSpeckByBasePath.getOrPut(path.basePath().referenceName) { mutableListOf() }.add(useSpeck)
+                }
+            }
+        }
+        return useSpeckByBasePath
+    }
+
+    private fun makeGroupedPath(useSpecksByBasePath: Map<String, List<RsUseSpeck>>): List<String> {
+        return useSpecksByBasePath.entries.map { (basePath, specks) ->
+            if (specks.size == 1) {  // Cannot group use specks with this $basePath
+                val useSpeck = specks[0]
+                useSpeck.text
+            } else { // Group use specks with this $basePath
+                val useSpecksInGroup = specks.flatMap {
+                    // Remove first group
+                    if (it.path?.text == basePath && it.useGroup != null) {
+                        it.useGroup!!.useSpeckList.map { it.text }
+                    } else {
+                        listOf(deleteBasePath(it.text, basePath))
+                    }
+                }
+                useSpecksInGroup.joinToString(",\n", "$basePath::{\n", ",\n}")
```

Comment:
Maybe we shouldn't insert a trailing comma?

Reply:
I'll change it to add a comma when it is needed.